### ISSUE TITLE
Fast-RTPS not acknowledging HEARTBEAT messages [4339]

### ISSUE THREAD

**Issue by neeleym** [eProsima/Fast-RTPS#375]
In conducting interoperability testing with the ShapesDemo and OpenDDS we see attempts of HEARTBEAT messages from OpenDDS not being acknowledged by the Fast-RTPS shapes client. The spec states that this should be implemented in 8.4.2.3.1:
```
Readers must respond eventually after receiving a HEARTBEAT with final flag not set
8.4.2.3.1
Upon receiving a HEARTBEAT Message with final flag not set, the Reader must respond with an ACKNACK Message.
The ACKNACK Message may acknowledge having received all the data samples or may indicate that some data samples
are missing.
```
Because of the above , we are not getting proper endpoint handshaking completed and, thus, no participant data exchange can follow.  Are you responding to HEARTBEAT messages per 8.4.2.3.1 above or do you have other configuration suggestions for subsequent testing?

Thanks,
Marc

Adding to above ...  The following link is how to run the OpenDDS shapes demo should this need to be reproduced ...
 [http://opendds.org/quickstart/GettingStartedShapesDemo.html](url)

**Comment by MiguelCompany** [comment]
Hi @neeleym 

Looking at the code I don't see any issue. The lines [here](https://github.com/eProsima/Fast-RTPS/blob/3a1bf2ab6670cf5295eae3af5641522c84716c35/src/cpp/rtps/reader/StatefulReader.cpp#L372-L375) schedule the heartbeat response timer when received final flag is false. Then the event code checks [here](https://github.com/eProsima/Fast-RTPS/blob/3a1bf2ab6670cf5295eae3af5641522c84716c35/src/cpp/rtps/reader/timedevent/HeartbeatResponseDelay.cpp#L74) if the last received heartbeat had a false final flag.

Moreover, Fast-RTPS is sending this kind of heartbeat when a reader is matched.

I think the only reasons for not sending the acknack back would be
* The reader is best-effort (and thus ignores heartbeats as 8.4.2.3 indicates)
* The heartbeat is discarded [here](https://github.com/eProsima/Fast-RTPS/blob/3a1bf2ab6670cf5295eae3af5641522c84716c35/src/cpp/rtps/messages/MessageReceiver.cpp#L809)
* The reader has not matched with the writer, so the message is discarded [here](https://github.com/eProsima/Fast-RTPS/blob/3a1bf2ab6670cf5295eae3af5641522c84716c35/src/cpp/rtps/reader/StatefulReader.cpp#L425)

**Comment by neeleym** [comment]
@MiguelCompany  ... you mention that "the reader is best-effort".  All of the OpenDDS RTPS traffic is reliable. Is the default for Fast-RTPS best-effort or defaulted to reliable? Do we need to build it differently to assure reliable at the RTPS level?

**Comment by MiguelCompany** [comment]
@neeleym You mentioned you are using ShapesDemo. When you click on Subscribe, there is an option for Reliable under QoS Settings that is disabled by default. Just ensure it is enabled and everything should work.

**Comment by mitza-oci** [comment]
The Heartbeat submessages we are referring to belong to the SEDP built-in entities, which are always reliable.

**Comment by MiguelCompany** [comment]
Hi @mitza-oci and @neeleym 

Have you checked with other DDS implementations?

**Comment by mitza-oci** [comment]
OpenDDS is successfully interoperating with all known DDS implementations.  We have demonstrated interoperability at multiple OMG meetings and are using OpenDDS in interop scenarios in production projects.

**Comment by MiguelCompany** [comment]
We have recently solved some discovery related issues which are going to be part of release 1.7.1 (#393). Could you check against the code there?

**Comment by MiguelCompany** [comment]
Hi @neeleym and @mitza-oci 

I finally discovered the problem, and have a fix ready on #414.

**Comment by neeleym** [comment]
That's good news @MiguelCompany .  I see it is in review. Do you have an estimation of when this will be merged into a release?

Thanks,
Marc

**Comment by MiguelCompany** [comment]
Hi @neeleym 

PR #414 has been merged into the 1.7.1 release branch. We plan to merge the release (#393) into master by the end of next week.
